app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTelemetry.kt

```diff
@@ -0,0 +1,159 @@
+        onCreateData = AppAllStartup(
+            getStartupSourceFromIntent(safeIntent, true),
+            getAppStartupType(),
+            hasSavedInstanceState
+        )
+        wasAppCreateCalledBeforeActivityCreate = false
+    }
+
+    /**
+     * we are not setting [Source] in this method since source is derived from a [SafeIntent].
+     * therefore source either gets set in oncreate() or onNewIntent().
+     */
+    fun onHomeActivityOnRestart() {
+        onRestartData = Pair(HOT, false)
+    }
+
+    /**
+     * we are only setting [Source] in this method since source is derived from a [SafeIntent].
+     * other metric fields are set in onCreate() and onRestart()
+     */
+    fun onHomeActivityOnNewIntent(safeIntent: SafeIntent) {
+        onNewIntentData = getStartupSourceFromIntent(safeIntent, false)
+    }
+
+    private fun getAppStartupType(): Type {
+        return if (wasAppCreateCalledBeforeActivityCreate) COLD else WARM
+    }
+
+    private fun getStartupSourceFromIntent(
+        intent: SafeIntent,
+        isExternalAppBrowserActivity: Boolean
+    ): Source {
+        return when {
+            isExternalAppBrowserActivity -> CUSTOM_TAB
+            intent.isLauncherIntent -> APP_ICON
+            intent.action == Intent.ACTION_VIEW -> LINK
+            else -> UNKNOWN
+        }
+    }
+
+    /**
+     * The reason we record metric on resume is because we need to wait for onNewIntent(), and
+     * we are not guaranteed that onNewIntent() will be called before or after onStart() / onRestart().
+     * However we are guaranteed onResume() will be called after onNewIntent() and  onStart(). Source:
+     * https://developer.android.com/reference/android/app/Activity#onNewIntent(android.content.Intent)
+     *
+     */
+    fun onHomeActivityOnResume() {
+        recordMetric()
+    }
+
+    private fun recordMetric() {
+        if (!isMetricRecordedSinceAppWasForegrounded) {
+            val appAllStartup: AppAllStartup = if (onCreateData != null) {
+                    onCreateData!!
+            } else if (onRestartData != null && onNewIntentData != null) {
+                    mergeOnRestartAndOnNewIntentIntoStartup()
+            } else {
+                getErrorStartup()
+            }
+            metrics.track(appAllStartup)
+            isMetricRecordedSinceAppWasForegrounded = true
+        }
+        // we don't want any weird previous states to persist on our next metric record.
+        onCreateData = null
+        onNewIntentData = null
+        onRestartData = null
+    }
+
+    private fun mergeOnRestartAndOnNewIntentIntoStartup(): AppAllStartup {
+        return AppAllStartup(
+            onNewIntentData!!,
+            onRestartData!!.first,
+            onRestartData!!.second
+        )
+    }
+    private fun getErrorStartup(): AppAllStartup {
+        return if (onNewIntentData == null && onRestartData == null) {
```

Comment:
Based on our understanding of the lifecycle, what are the impossible cases?

| onNewIntent | onRestart | Type | Source |
| - | - | - | - |
| data | data | HOT | from Intent |
| null | data | HOT | UNKNOWN (possibly app switcher) |
| data | null | ERROR | from Intent |
| null | null | ERROR | UNKNOWN |

The error cases occur because we never the lifecycle methods to get called in that order for it to happen, right?
- In the 3rd row, this case is we didn't create the activity (onCreate) and we were backgrounded but `onRestart` wasn't called: we don't expect that to ever happen, right?
- In the 4th row, this case is we didn't create the activity and we were backgrounded but neither `onRestart` or `onNewIntent` was called: we don't expect that to happen either, right?

Generally, the assumption is `onRestart` should always be called if we're backgrounded and `onCreate` wasn't called.

Do you agree?

**I believe the code is wrong: can you update it to fit this model?** Note how I didn't duplicate error/unknown through type/source too.

Generating tables like these is helpful for ensuring you think through all possible states.